### PR TITLE
WT-7833 wt_btree_switch_object needs to coordinate with concurrent write requests to btree

### PR DESCRIPTION
**Pull request by keithbostic**

### CODE REVIEW

File: src/block/block_open.c
Comment (sueloverso) [inline src/block/block_open.c:135]: This line feels strange in this location splitting the two `wt_free` calls. Maybe alphabetize the function and put below the 2nd free, or alphabetize the field and move above the first free.

File: src/block_cache/block_mgr.c
Comment (sueloverso) [inline src/block_cache/block_mgr.c:153]: @keitharnoldsmith is this the line you mean for WT-9503?

**Review by sueloverso — APPROVED** [review]
I've done a first pass. LGTM. But I plan to re-read it to understand this more but approving now.

File: test/csuite/tiered_abort/main.c
Comment (keithbostic) [reply test/csuite/tiered_abort/main.c:138]: I don't know anything about those programs, so I'm leaving that to someone else.  Building with this branch does report:
```
####################### MESSAGE ############################
s_all run of: "./s_mentions" resulted in:
    There are comments mentioning wt-7833 in the code, please check if they need to be resolved:
    grep: input file '../dist/__wt.2763969' is also the output
    ../test/csuite/schema_abort/main.c:90: * TODO: WT-7833 Lock to coordinate inserts and flush_tier. This lock should be removed when that
    ../test/csuite/random_directio/main.c:168: * TODO: WT-7833 Lock to coordinate inserts and flush_tier. This lock should be removed when that
```

File: src/include/block.h
Comment (keithbostic) [reply src/include/block.h:243]: Fixed in d8765bc.

File: src/block/block_ckpt.c
Comment (keithbostic) [reply src/block/block_ckpt.c:78]: I figured it was a cheap enough check that we might reasonably always make it rather than limiting it to #DIAGNOSTIC.

I also switched to `WT_ERR_ASSERT` in 5e3604d.

File: src/block/block_open.c
Comment (keithbostic) [reply src/block/block_open.c:135]: This is in `WT_BLOCK` structure field order.

File: src/block_cache/block_mgr.c
Comment (keithbostic) [reply src/block_cache/block_mgr.c:153]: I'll add a comment to that ticket.

**Review by keitharnoldsmith — APPROVED** [review]
LGTM.  One comment on a comment :-)

**Comment by agorrod** [comment]
@ddanderson or @sueloverso would it make sense for one of you to take over ownership of this PR and get it ready for merge?

**Comment by agorrod** [comment]
ping @sueloverso and @ddanderson

**Comment by sueloverso** [comment]
I am going to defer this to @ddanderson because he had a similar PR and had his head in this space and can better judge this PR.

**Comment by ddanderson** [comment]
@agorrod, yes I'll take over this PR.  I don't know if there's any way to make me the owner, but I'll move it.

**Comment by ddanderson** [comment]
@sueloverso and @keitharnoldsmith, I know we've all looked at this before, but please take chance to review before it's merged.  I didn't change anything other than merging the latest develop into it.  @keitharnoldsmith, would you care to update the comment you had an issue with?  Thanks!

**Comment by ddanderson** [comment]
@sueloverso, I removed the extra locking in test programs in the latest commit.

**Review by sueloverso — APPROVED** [review]
Thank you @ddanderson. LGTM

**Review by ddanderson — APPROVED** [review]
LGTM

**Comment by ddanderson** [comment]
@keitharnoldsmith, I fixed the comment you had issue with.  I don't know if you want to give this another look before it gets merged?

File: src/block/block_ckpt.c
Comment (keitharnoldsmith) [inline src/block/block_ckpt.c:77]: I'm not sure of the intent of the two error checks here. Both are triggered if `live_open == true`, but we'll never get to the second. If compiled with `HAVE_DIAGNOSTIC`, `WT_ERR_ASSERT` will abort. Otherwise it calls `WT_ERR_MSG` which includes a `goto err` statement.

**Review by keitharnoldsmith — APPROVED** [review]
LGTM.  I commented on one place where the error handling could be cleaned up.

File: src/block/block_ckpt.c
Comment (ddanderson) [reply src/block/block_ckpt.c:77]: Good catch.  I'll remove the second of these, I like keeping the assert  in diagnostic mode.

Comment (keitharnoldsmith) [reply src/block/block_ckpt.c:77]: I think it would also be useful to capture `block->name" from the second message.  But no reason it can't be part of the 1st message.

**Comment by ddanderson** [comment]
@keitharnoldsmith, ping for one more look on adjusted PR.

**Review by keitharnoldsmith — APPROVED** [review]
LGTM!

**Pull request merged**
